scene/gui/box_container.cpp

```diff
@@ -238,6 +241,8 @@ void BoxContainer::_resort() {
 		}
 
 		fit_child_in_rect(c, rect);
+		cached_children_pos.push_back(vertical ? rect.position.y : rect.position.x);
+		cached_children_idx.push_back(i);
```

Comment:
When a non-control is added to a container, the indices of controls may change, but no resorting is needed, so `_resort()` won't run. So the logic will break because the indices are now wrong, likewise with other containers. I think you can remedy this by putting the controls directly inside the array.

No idea how this idea would affect performance.

Reply:
That's indeed a problem:

https://github.com/godotengine/godot/assets/6501975/6ecd969d-2986-4b4d-9398-ff5780ad4121

> I think you can remedy this by putting the controls directly inside the array and modifying the rest of the logic slightly.

I think caching the nodes directly will only work for `BoxContainer`s, since they only have to return one child. For the other containers, I need to return a list of nodes, and through a node reference I don't see a good way to do this. I could use `get_children()` to find where that node exists and then iterate over it, but beyond incurring that search cost, this method performs a copy of the node data, which looks extra wasteful: https://github.com/godotengine/godot/blob/3a8524dd923e9d9a79e4979a47fd427388ea7010/scene/main/node.cpp#L1548-L1558

Reply:
In GridContainer, you can get the row and column from division/modulo, it should be similar to BoxContainer but done twice. For FlowContainer, you could store the number of controls in each row/column, which won't change in this situation, unlike indices.

Reply:
Those are great ideas, I'll try them in the morning!

Reply:
Alright, so after thinking on how to implement these, I noticed that the issue would remain: while those would fix the problem of how to obtain the next nodes after the starting one, I need a way to know where the starting node lies in the children array; therefore, its index. Turns out, though, that it is possible to do just that with `Node::get_index()`, which I had no idea it existed. 

With that in mind, it's trivial to keep the current system, since now I'll have an array of cached nodes where I can obtain their updated indices.

Reply:
Awesome, so you'll only need one array then and it's a code simplification if anything. I'm very glad I caught this now, instead of requiring an extensive investigation of why buttons don't work in a fringe case, years down the line.

Reply:
Pushed new changes, now this issue should no longer occur. Changes in speedups were also negligible (averaging to 0.18% faster)

Reply:
Do you really need to cache the positions if you have the nodes? Just wondering

Reply:
That's a good point, now that there's direct node access there isn't much point in keeping a separate position array. I'll try this in a bit :+1:

Reply:
Pushed now. I still haven't tested changes in speedups on all scenarios, but from quick testing, it doesn't seem to be significantly impactful. Will have to update the whole table since my baseline seems to be different now though.

Reply:
Updated benchmark data, and it seems it did improve a bit, averaging +5.629%.